lib/utility.js

```diff
@@ -1,10 +1,171 @@
 import Product from "../models/Product";
 
-// NOTE: variables and functions whose name starts with "_" are private and not to be exported
+// const marketUrl = "http://testcusa-env.afejb3cmmd.us-east-2.elasticbeanstalk.com/MarketExecute.php";
+// const loginUrl = "http://testcusa-env.afejb3cmmd.us-east-2.elasticbeanstalk.com/LoginExecute.php"
+const marketUrl = "https://madishare.com/MarketExecute.php";
+const loginUrl = "https://madishare.com/LoginExecute.php";
+let app = getApp();
 
-const _marketUrl = "http://testcusa-env.afejb3cmmd.us-east-2.elasticbeanstalk.com/MarketExecute.php";
+/** @review
+ * Local function to promisify wx request 
+ * @param endpoint: url of the endpoint being requested
+ * @param data: POST data field for the endpoint
+ * @param filePath: path of the file attached to the form (OPTIONAL)
+ * @promise resolve: the response data if request completes successfully,
+ * @promise reject: if either ErrorCode is set or wx fail to send the request
+ */
+function requestPromise(endpoint, data, filePath){
+  let hasFile = filePath!=undefined;
+  let req = hasFile?wx.uploadFile:wx.request;
+  return new Promise(function(resolve, reject){
+    let config = { url: endpoint,
+      success: (res) => { 
+        if (res.hasOwnProperty('ErrorCode') && res.ErrorCode != 1)  // Check ErrorCode if exist
+            reject({ ErrorCode: res.ErrorCode,
+                     ErrorMessage: res.ErrorMessage });
+        resolve(res.data);  
+      },
+      fail: (res) => { reject (res); } 
+    }
+    if (!hasFile) {
+      config['method'] = "POST";
+      config['header'] = { "content-type": "application/x-www-form-urlencoded", };
+      config['dataType'] = "json";
+      config['data'] = data;
+    }else{
+      config['filePath'] = filePath;
+      config['formData'] = data;
+      config['name'] = "ProductImage";
+    }
+    req(config);
+  });
+}
 
 /**
+ * Choose images (Max: 9)
+ * @promise resolve: the tmp imgPaths (Array[String])
+ * @promise reject: Fail to extract images
+ */
+function chooseImagePromise(){
+  let config = {
+    "count": 9, // at-most 9 images at onces selected
+    "sizeType": "original",
+    "sourceType": "album"
+  }
+  return new Promise(function(resolve, reject){
+    config.success = res => {resolve(res.tempFilePaths)};
+    config.fail = err => {reject(err)};
+    wx.chooseImage(config);
+  })
+}
+
+/** @review
+ * Promisified function for user login
+ * Also can be used to check existence of globalData.userInfo
+ * @param config (Optional): the globalData object used to automatically store the retrieved UserID 
+ * @promise resolve: UserId (unique) from server
+ * @promise reject:  Fail to generate wx code or fail to retrieve UserId from server
+ */
+function getUserId(globalData){ 
+  return new Promise(function(resolve, reject){

```

Comment:
Instead of creating and returning a new promise that initiates a new login server request every time, I think it would be good to have a global singleton promise variable (sth like "_loginConnection"). Then this function just instantiate/return this singleton on every request.

This advantages are:
1) Less number of requests
2) When login finishes, all callback waiting on this are fired at the same time. If we return a new promise every single time, one user might wait on the promise it gets when another promise is already returned.